File: src/pokeevaluator/core/stats.py
"""Gen III stat formulas and IV recovery."""

from __future__ import annotations

import math

from pokeevaluator.data.gen3 import DEFAULT_EV, IV_RANGE
from pokeevaluator.models.pokemon import BaseStats, StatBlock, StatName, STAT_ORDER
from pokeevaluator.models.specimen import IVRange, IVSet


def calc_hp(base: int, iv: int, level: int, ev: int = DEFAULT_EV) -> int:
    """Gen III HP formula: floor((2*Base + IV + floor(EV/4)) * Level / 100) + Level + 10."""
    return math.floor((2 * base + iv + math.floor(ev / 4)) * level / 100) + level + 10


def calc_other_stat(
    base: int, iv: int, level: int, nature_mod: float, ev: int = DEFAULT_EV
) -> int:
    """Gen III non-HP stat formula: floor((floor((2*Base + IV + floor(EV/4)) * Level / 100) + 5) * NatureMod)."""
    inner = math.floor((2 * base + iv + math.floor(ev / 4)) * level / 100) + 5
    return math.floor(inner * nature_mod)


def recover_iv_hp(base: int, level: int, observed: int, ev: int = DEFAULT_EV) -> IVRange:
    """Recover possible HP IVs by brute force."""
    candidates = [iv for iv in IV_RANGE if calc_hp(base, iv, level, ev) == observed]
    if not candidates:
        raise ValueError(f"Nenhum IV possível para HP observado={observed}")
    return IVRange(stat=StatName.HP, min_iv=candidates[0], max_iv=candidates[-1])


def recover_iv_other(
    stat: StatName,
    base: int,
    level: int,
    nature_mod: float,
    observed: int,
    ev: int = DEFAULT_EV,
) -> IVRange:
    """Recover possible IVs for a non-HP stat by brute force."""
    candidates = [
        iv
        for iv in IV_RANGE
        if calc_other_stat(base, iv, level, nature_mod, ev) == observed
    ]
    if not candidates:
        raise ValueError(f"Nenhum IV possível para {stat.value} observado={observed}")
    return IVRange(stat=stat, min_iv=candidates[0], max_iv=candidates[-1])


def recover_all_ivs(
    base_stats: BaseStats,
    level: int,
    nature_mods: StatBlock,
    observed: StatBlock,
) -> IVSet:
    """Recover IVs for all 6 stats."""
    hp_range = recover_iv_hp(
        base=base_stats.hp, level=level, observed=int(observed.hp)
    )

    other_ranges: dict[StatName, IVRange] = {}
    for stat in STAT_ORDER:
        if stat == StatName.HP:
            continue
        other_ranges[stat] = recover_iv_other(
            stat=stat,
            base=base_stats[stat],
            level=level,
            nature_mod=nature_mods[stat],
            observed=int(observed[stat]),
        )

    return IVSet(
        hp=hp_range,
        atk=other_ranges[StatName.ATK],
        def_=other_ranges[StatName.DEF],
        spatk=other_ranges[StatName.SPATK],
        spdef=other_ranges[StatName.SPDEF],
        spe=other_ranges[StatName.SPE],
    )

File: web/src/lib/core/stats.ts
import { BaseStats, IVRange, IVSet, StatBlock, StatName, STAT_ORDER, getStatFromBaseStats, getStatFromBlock } from "../types";
import { DEFAULT_EV, MAX_IV, MIN_IV } from "../data/gen3";

export function calcHp(base: number, iv: number, level: number, ev: number = DEFAULT_EV): number {
  return Math.floor((2 * base + iv + Math.floor(ev / 4)) * level / 100) + level + 10;
}

export function calcOtherStat(
  base: number, iv: number, level: number, natureMod: number, ev: number = DEFAULT_EV
): number {
  const inner = Math.floor((2 * base + iv + Math.floor(ev / 4)) * level / 100) + 5;
  return Math.floor(inner * natureMod);
}

export function recoverIvHp(base: number, level: number, observed: number, ev: number = DEFAULT_EV): IVRange {
  const candidates: number[] = [];
  for (let iv = MIN_IV; iv <= MAX_IV; iv++) {
    if (calcHp(base, iv, level, ev) === observed) {
      candidates.push(iv);
    }
  }
  if (candidates.length === 0) {
    throw new Error(`No possible IV for observed HP=${observed}`);
  }
  return { stat: StatName.HP, minIv: candidates[0], maxIv: candidates[candidates.length - 1] };
}

export function recoverIvOther(
  stat: StatName, base: number, level: number, natureMod: number, observed: number, ev: number = DEFAULT_EV
): IVRange {
  const candidates: number[] = [];
  for (let iv = MIN_IV; iv <= MAX_IV; iv++) {
    if (calcOtherStat(base, iv, level, natureMod, ev) === observed) {
      candidates.push(iv);
    }
  }
  if (candidates.length === 0) {
    throw new Error(`No possible IV for ${stat} observed=${observed}`);
  }
  return { stat, minIv: candidates[0], maxIv: candidates[candidates.length - 1] };
}

export function recoverAllIvs(
  baseStats: BaseStats, level: number, natureMods: StatBlock, observed: StatBlock
): IVSet {
  const hpRange = recoverIvHp(baseStats.hp, level, observed.hp);

  const otherRanges: Partial<Record<StatName, IVRange>> = {};
  for (const stat of STAT_ORDER) {
    if (stat === StatName.HP) continue;
    otherRanges[stat] = recoverIvOther(
      stat,
      getStatFromBaseStats(baseStats, stat),
      level,
      getStatFromBlock(natureMods, stat),
      getStatFromBlock(observed, stat),
    );
  }

  return {
    hp: hpRange,
    atk: otherRanges[StatName.ATK]!,
    def_: otherRanges[StatName.DEF]!,
    spatk: otherRanges[StatName.SPATK]!,
    spdef: otherRanges[StatName.SPDEF]!,
    spe: otherRanges[StatName.SPE]!,
  };
}
